spec/unit/deep_preloader_spec.rb

```diff
@@ -235,6 +235,62 @@
     end
   end
 
+  context 'with a has_many STI relationship' do
+    with_temporary_table(:parent) do
+      has_many :pets, inverse_of: :parent
+    end
+
+    with_temporary_table(:pet, ->(t) { t.string :type; t.references :parent; t.references :cat_toy; t.references :dog_toy }) do
+      belongs_to :parent, inverse_of: :pets
+    end
+
+    with_temporary_table(:cat, superclass: -> { Pet }, create_table: false) do
+      belongs_to :cat_toy
+    end
+
+    with_temporary_table(:dog, superclass: -> { Pet }, create_table: false) do
+      belongs_to :dog_toy
+    end
+
+    with_temporary_table(:cat_toy) do
+      has_one :cat, inverse_of: :cat_toy
+    end
+
+    with_temporary_table(:dog_toy) do
+      has_one :dog, inverse_of: :dog_toy
+    end
+
+    let(:parent) do
+      Parent.create!(pets: pets)
+    end
+
+    let(:pets) do
+      [Dog.create(dog_toy: DogToy.new), Cat.create(cat_toy: CatToy.new)]
+    end
+
+    before(:each) do
+      parent.reload
+    end
+
+    it 'loads the children' do
+      DeepPreloader.preload(parent, :pets)
+      expect(parent).to have_loaded(:pets).as(pets)
+    end
+
+    it 'loads polymorphically through the children' do
+      DeepPreloader.preload(parent, :pets => DeepPreloader::PolymorphicSpec.parse(Dog.name => :dog_toy, Cat.name => :cat_toy))
+      expect(parent).to have_loaded(:pets).as(pets)
+      expect(pets[0]).to have_loaded(:parent).as(parent)
+      expect(pets[1]).to have_loaded(:parent).as(parent)
+
+      expect(parent.pets[0]).to have_loaded(:dog_toy).as(pets[0].dog_toy)
```

Comment:
I don't see an ordering constraint that will always cause the dog to be the first pet returned.

```suggestion
      expect(parent.pets.detect { |pet| pet.is_a?(Dog) }).to have_loaded(:dog_toy).as(pets[0].dog_toy)
```

Though looking at our other `has_many` test, it also seems to rely on insertion order being returned and hasn't been a problem, so maybe we don't have to worry about this now.

Reply:
Yeah, it's undefined but happens to be the case. Perhaps I should `order(:id)` just to make sure it's well-defined.